### PR TITLE
Introduce solve[_from_local]! and deprecate non-!

### LINKED ISSUE #17: Rename `solve` to `solve!`?
It looks like `solve` mutate the input:

```julia
julia> ex = :(x = 1)
       solve(ex)
       ex
:(@global x = 1)
```

If this is intentional, I think it'd be nice to call it `solve!` to be compatible with the usual convention (e.g., `fill!`).  If not, maybe `deepcopy` the input before doing the analysis?

### PR DESCRIPTION
**Pull request by tkf**
close #17

### CODE REVIEW

File: src/JuliaVariables.jl [hunk 1]
```diff
@@ -504,6 +504,23 @@ function solve(ast; toplevel=true)
     transform(ast)
 end # module struct
 
-solve_from_local(@nospecialize(ex)) = solve(ex; toplevel=false)
+solve_from_local!(@nospecialize(ex)) = solve!(ex; toplevel=false)
+
+_depwarn(sym) = Base.depwarn(
```
Comment (thautwarm) [inline src/JuliaVariables.jl:509]: Oh great!

File: test/runtests.jl [hunk 2]
```diff
@@ -389,9 +389,29 @@ end
...
         @inline function ()
         end
-    end) |> solve_from_local
+    end) |> solve_from_local!
     tp = first_scope_except_top(ast)
     scope = tp[1]
     @test isempty(scope.bounds)
     @test isempty(scope.freevars)
 end
+
```
Comment (thautwarm) [inline test/runtests.jl:398]: Thanks for adding new test sets!
Besides, I'd ask if you are using refactor tools to rename the `solve` to `solve!`?

File: test/runtests.jl [hunk 3]
```diff
@@ -128,20 +128,20 @@ get_fn_body(ex) =
...
-    a = solve(:(function z(x, k=1)
+    a = solve!(:(function z(x, k=1)
                    (k=k, )
                end
     ))
 
     @test any(x -> x.name == :k, get_fn_scope(a).bounds)
 
-    @test @when :(k=$_, ) = get_fn_body(a).args[2] begin
+    @test @when :(k=$_, ) = get_fn_body(a).args[end] begin
```
Comment (tkf) [inline test/runtests.jl:138]: I needed to change `.args[2]` to `.args[end]` to make it work with Julia 1.5-DEV.  It seems the AST was changed a bit in Julia 1.5-DEV:

```julia
julia> :(function f()
           a
       end)
:(function f()
      #= REPL[1]:1 =#
      #= REPL[1]:2 =#
      a
  end)
```

while it is

```julia
julia> :(function f()
           a
       end)
:(function f()
      #= REPL[1]:2 =#
      a
  end)
```

in Julia 1.4 and before.

**Review by thautwarm — APPROVED** [review]

**Comment by thautwarm** [comment]
I'd appreciate a lot for this PR!

File: src/JuliaVariables.jl [hunk 4]
```diff
@@ -504,6 +504,23 @@ function solve(ast; toplevel=true)
...
+    "`$sym(ex)` is deprecated in JuliaVariables 0.2.x. " *
+    "Please use `$(sym)!(ex)` instead.  `$sym(ex)`, which currently mutates `ex` " *
+    "in-place, will be re-introduced in 0.3 as a non-mutating variant.",
+    sym,
+)
```
Comment (tkf) [inline src/JuliaVariables.jl:514]: I wasn't sure if you want to go straight to 0.3 or stay at 0.2 a bit more.  But I guess deprecating `solve` for now while introducing `solve!` would be the most user-friendly thing to do at the moment?

Comment (thautwarm) [reply src/JuliaVariables.jl:514]: I won't go to 0.3 immediately, but it wouldn't take a long time.
> But I guess deprecating solve for now while introducing solve! would be the most user-friendly thing to do at the moment?

Right.

File: test/runtests.jl (hunk 2 above)
Comment (tkf) [reply test/runtests.jl:398]: Well, it was just a simple search-and-replace (`%s/\<solve\>/\0!/g` in vim).  Sorry, it was a super boring technique :)

**Comment by tkf** [comment]
Thanks a lot for the super quick review+merge!

File: test/runtests.jl (hunk 2 above)
Comment (thautwarm) [reply test/runtests.jl:398]: haha, it works! thank you!

**Pull request merged**